### PR TITLE
[search] Better streets ranking and matching.

### PR DESCRIPTION
**Pull request by vng**
Street + House Number match happens only with correctly mapped (OSM addr:street + addr:housenumber) addresses since now.

Pros:
- Speed
- Can increase kLookupRadiusM to fix some fancy cases like here https://github.com/organicmaps/organicmaps/issues/5971
- Prepare for addr:place (also only exact mapping)

Cons:
- Can't find some bad OSM-mapped addresses, but they may be ambiguous.

This approach is good since we are going to integrate other open address sources (like https://openaddresses.io/)

### CODE REVIEW

File: search/ranker.cpp [hunk 1]
```diff
@@ -115,8 +123,20 @@ NameScores GetNameScores(FeatureType & ft, Geocoder::Params const & params,
...
       if (type == Model::TYPE_STREET)
       {
+        // Searching for "Santa Fe" should rank "Avenida Satna Fe" like FULL_MATCH or FULL_PREFIX, but not SUBSTRING.
+        {
+          TokensVector cleaned(RemoveStreetSynonyms(vec.GetTokens()));
+          UpdateNameScores(cleaned, lang, slice, bestScores);
+          UpdateNameScores(cleaned, lang, sliceNoCategories, bestScores);
+        }
+
+        /// @todo Should definitely add "platz"-"pl". And maybe "gasse"-"g", "allee"-"al"?
```
Comment (biodranik) [inline search/ranker.cpp]:
```
✗ rg -iwc 'str\.' all.txt       
13863

✗ rg -iwc 'pl\.' sorted.txt 
1451

✗ rg -iwc ' g\.' sorted.txt
1217


✗ rg -iwc ул\. all.txt
32450

✗ rg -iwc пл\. all.txt
2600

✗ rg -iwc пер\. all.txt
5601

```

Some interesting addresses (check yourself):
```
E.-Claudius-Str./H.-Mann-Allee
Bernauer Str./Wassersportvereine
```

https://cdn.organicmaps.app/apk/unique_osm_names.zip

File: indexer/search_string_utils.cpp [hunk 2]
```diff
@@ -498,4 +498,14 @@ void StreetTokensFilter::Put(UniString const & token, bool isPrefix, size_t tag)
 
   m_callback(token, tag);
 }
+
+std::map<UniString, UniString> const & GetDACHStreets()
+{
+  static std::map<UniString, UniString> res = {
+    { MakeUniString("strasse"), MakeUniString("str") },
+    { MakeUniString("platz"), MakeUniString("pl") },
```
Comment (biodranik) [inline indexer/search_string_utils.cpp:506]: Is `pl.` converted to `pl`?

File: indexer/search_string_utils.hpp [hunk 3]
```diff
@@ -128,4 +128,8 @@ class StreetTokensFilter
   Callback m_callback;
   bool m_withMisprints = false;
 };
+
+// D-A-CH countries has special street suffixes processing.
```
Comment (biodranik) [inline indexer/search_string_utils.hpp:132]: s/has/have/

File: indexer/search_string_utils.hpp [hunk 4]
```diff
@@ -128,4 +128,8 @@ class StreetTokensFilter
...
+std::map<strings::UniString, strings::UniString> const & GetDACHStreets();
```
Comment (biodranik) [inline indexer/search_string_utils.hpp:133]: typedef it?

File: search/ranker.cpp [hunk 5]
```diff
@@ -54,41 +54,45 @@ void UpdateNameScores(TokensVector & tokens, uint8_t lang, Slice const & slice,
...
+
+      // "Abcd strasse" -> "abcdstr".
+      result.emplace_back(streetTokens.begin(), streetTokens.end() - 1);
+      result.back().back() += sx.second;
+      return result;
+    }
+
+    // "Abcdstrasse" -> "abcd strasse".
+    auto const name =
+        strings::UniString(streetTokens.back().begin(), streetTokens.back().end() - sx.first.size());
```
Comment (biodranik) [inline search/ranker.cpp:86]: One line?

File: search/street_vicinity_loader.cpp [hunk 6]
```diff
@@ -47,11 +47,11 @@ void StreetVicinityLoader::LoadStreet(uint32_t featureId, Street & street)
   if (!feature)
     return;
 
-  bool const isStreet = feature->GetGeomType() == feature::GeomType::Line &&
-                        ftypes::IsWayChecker::Instance()(*feature);
-  bool const isSquareOrSuburb = ftypes::IsSquareChecker::Instance()(*feature) ||
-                                ftypes::IsSuburbChecker::Instance()(*feature);
-  if (!isStreet && !isSquareOrSuburb)
+  bool const isStreet =
+      (feature->GetGeomType() == feature::GeomType::Line &&
```
Comment (biodranik) [inline search/street_vicinity_loader.cpp:51]: Не люблю такое форматирование, на одну строку больше получается.

**Review by biodranik — APPROVED** [review]

File: search/ranker.cpp (hunk 1 above)
Comment (vng) [reply search/ranker.cpp]: The patch is for German-lang addresses that have long, short and concatenated prefix.
strasse, platz is enough

File: indexer/search_string_utils.cpp (hunk 2 above)
Comment (vng) [reply indexer/search_string_utils.cpp:506]: yes, not here

File: search/ranker.cpp [hunk 7]
```diff
@@ -115,8 +123,20 @@ NameScores GetNameScores(FeatureType & ft, Geocoder::Params const & params,
...
 
+      /// @todo
+      /// 1. Make sure that this conversion also happens for Address and POI results,
+      /// where street is only one component.
+      /// 2. Make an optimization: If there are no synonyms or "strasse", skip this step.
       if (type == Model::TYPE_STREET)
       {
+        // Searching for "Santa Fe" should rank "Avenida Satna Fe" like FULL_MATCH or FULL_PREFIX, but not SUBSTRING.
+        {
+          TokensVector cleaned(RemoveStreetSynonyms(vec.GetTokens()));
```
[the remark this answers is not in the document]
Comment (vng) [reply search/ranker.cpp:138]: Function takes but ref below.

**Comment by vng** [comment]
@RicoElectrico 
- Searching for "Santa Fe" should give "Avenida Santa Fe" nearby street first. Test in Buenos Aires, where there are a lot of different "Santa Fe" staff. No generator needed.
- Searching for "Bahnhofpl" should match "Bahnhof platz" street. Check for example in Bern. Generator needed.

**Pull request merged**